File: Piece.java
public class Piece {
    long[] bitmaps; 
    int[][] coords;
    char rep;
    int[] heights, widths;
    int height, width;
    int nCoords;
    int nOrientations;

    public Piece(int[][] coords, char rep){
        this.coords = coords;
        this.rep = rep;
        this.nCoords = coords.length;
        this.bitmaps = new long[8];
        this.heights = new int[8];
        this.widths = new int[8];
        this.getOrientations();
    }

    public long createBitmap() {
        long bits = 0;
        for (int i = 0; i < 8; i++){
            for (int j = 0; j < 8; j++){
                bits <<= 1;
                for (int k = 0; k < nCoords; k++) {
                    if (coords[k][0] == i & coords[k][1] == j){
                        bits++;
                    }
                }
            }
        }
        return bits;
    }

    public void updateDims(){
        height = 0;
        width = 0;
        for (int i = 0; i < nCoords; i++) {
            if (height < coords[i][0]){
                height = coords[i][0];
            }
            if (width < coords[i][1]) {
                width = coords[i][1];
            }
        }
        height++;
        width++;
    }

    public int[][] rotate(){
        int[][] newCoords = new int[this.nCoords][2];
        int min_x = Integer.MAX_VALUE;
        int min_y = Integer.MAX_VALUE;

        for (int i = 0; i < this.nCoords; i++){
            newCoords[i][0] = -this.coords[i][1];
            if (min_x > newCoords[i][0]){
                min_x = newCoords[i][0];
            }

            newCoords[i][1] = this.coords[i][0];
            if (min_y > newCoords[i][1]){
                min_y = newCoords[i][1];
            }
        }

        for (int i = 0; i < this.nCoords; i++) {
            newCoords[i][0] -= min_x;
            newCoords[i][1] -= min_y;
        }

        return newCoords;
    }

    public int[][] flip(){
        int[][] newCoords = new int[this.nCoords][2];
        int min_x = Integer.MAX_VALUE;

        for (int i = 0; i < nCoords; i++) {
            newCoords[i][0] = -coords[i][0];
            newCoords[i][1] = coords[i][1];
            if (newCoords[i][0] < min_x){
                min_x = newCoords[i][0];
            }
        }

        for (int i = 0; i < nCoords; i++) {
            newCoords[i][0] -= min_x;
        }
        return newCoords;
    }

    public void rotateThis(){
        this.coords = this.rotate();
    }

    public void flipThis(){
        this.coords = this.flip();
    }

    public void getOrientations(){
        long bitmap;
        Boolean flag;
        this.updateDims();
        nOrientations = 0;
        // first orientation
        bitmaps[nOrientations] = this.createBitmap();
        heights[nOrientations] = height;
        widths[nOrientations] = width;
        nOrientations++;
        
        for (int i = 0; i < 3; i++){
            this.rotateThis();
            bitmap = this.createBitmap();
            flag = Boolean.FALSE;
            for (long l : bitmaps) {
                if (l == bitmap) {
                    flag = Boolean.TRUE;
                }
            }
            if (!flag){
                this.updateDims();
                bitmaps[nOrientations] = this.createBitmap();
                heights[nOrientations] = height;
                widths[nOrientations] = width;
                nOrientations++;
            }
        }

        this.flipThis();
        for (int i = 0; i < 4; i++){
            this.rotateThis();
            bitmap = this.createBitmap();
            flag = Boolean.FALSE;
            for (long l : bitmaps) {
                if (l == bitmap) {
                    flag = Boolean.TRUE;
                }
            }
            if (!flag){
                this.updateDims();
                bitmaps[nOrientations] = this.createBitmap();
                heights[nOrientations] = height;
                widths[nOrientations] = width;
                nOrientations++;
            }
        }      
        
    }

    public void print(){
        for (int i = 0; i < nCoords; i++) {
            System.out.println(String.format("%d, %d", coords[i][0], coords[i][1]));
        }
    }

    public void printGrid(){
        String strout = "";
        boolean contains;
        for (int i = 0; i < 8; i++){
            for (int j = 0; j < 8; j++){
                contains = Boolean.FALSE;
                for (int k = 0; k < nCoords; k++) {
                    if (coords[k][0] == i & coords[k][1] == j){
                        contains = Boolean.TRUE;
                    }
                }
                if (contains){
                    strout += "X ";
                }
                else{
                    strout += "O ";
                }
            }
            strout += "\n";
        }
        System.out.println(strout);
    }
    
    public void printOrientations(){
        for (int k = 0; k < nOrientations; k++) {
            long bits = bitmaps[k];
            long bit;
            for (int i = 0; i < 8; i++) {
                for (int j = 0; j < 8; j++) {
                    bit = (bits >> (63 - (i * 8 + j))) & 1;
                    System.out.print(bit == 1 ? this.rep + " " : "O ");
                }
                System.out.println();
            }
            System.out.println();

        }
    }
}

File: piece.py

from orientation import *
# piece is an array of coordinates
# contains the coordinates for the eight rotations in self.orientations
# char is the character rep for the piece

class Piece:
    def __init__(self, coords, char = None):
        self.coords = set(coords)
        self.orientations = []
        self.uniqueOrientations = 0
        self.height, self.width = self.getDims(self.coords)
        self.dims = max(self.height, self.width)
        if char:
            self.char = char
        else:
            self.char = "X"
        self.createOrientations()

    def getDims(self, coords):
        height = 0
        width = 0
        for i,j in coords:
            if i > height:
                height = i
            if j > width:
                width = j
        height = height + 1
        width = width + 1
        return (height, width)

    def rotate(self, coords):
        newCoords = set()
        newx = []
        newy = []
        min_x = min_y = 10000
        for x,y in coords:
            newx.append(-y)
            if -y < min_x:
                min_x = -y
            newy.append(x)
            if x < min_y:
                min_y = x

        for i,j in zip(newx, newy):
            newCoords.add((i-min_x, j-min_y))
        return newCoords

    # code is scuffed but should only run once so it's fine?
    def createOrientations(self):
        allOrientations = [0] * 8
        allOrientations[0] = self.coords
        for i in range(3):
            allOrientations[i+1] = self.rotate(allOrientations[i])
        for i in range(4):
            allOrientations[i+4] = self.flip(allOrientations[i])
        
        coordlist = []
        for i in range(8):
            if not allOrientations[i] in coordlist:
                coordlist.append(allOrientations[i])
        
        for i in range(len(coordlist)):
            height, width = self.getDims(coordlist[i])
            self.orientations.append(Orientation(coordlist[i], height, width, self.char))
    
    def getOrientation(self, n):
        if n >= len(self.orientations):
            print("Orientation doesn't exist")
            return False
        return self.orientations[n]

    def getOrientations(self):
        return self.orientations
        
    def flip(self, coords):
        newCoords = set()
        newx = []
        newy = []
        min_x = 10000
        for x,y in coords:
            newx.append(-x)
            if -x < min_x:
                min_x = -x
            newy.append(y)
        
        for i,j in zip(newx, newy):
            newCoords.add((i-min_x, j))
        return newCoords
    
    def printOrientations(self):
        print()
        for rot in range(len(self.orientations)):
            print("Orientation " + str(rot + 1) + ": ")
            self.orientations[rot].print()
            print()

    # def __eq__(self, value):
    #     return self.coords == value.coords
    


